records/2024-01-22.reshaping-cicd-environments.md

```diff
@@ -0,0 +1,127 @@
+environment to experiment on new infrastructure changes in an isolated manner,
+increasing velocity with changes of this nature.
+
+This also comes at a time when there is a desire to increase automation of the
+release process among the team, potentially improving the release process and
+decreasing the time spent on the ops lead routine tasks. This has to be done
+with certain considerations in mind such as risk, testing, traceability,
+rollback and accountability.
+
+> <sup>[1]</sup> A business continuity plan (BCP) is a strategic playbook created to help an organization maintain or quickly resume business functions in the face of disruption, whether that disruption is caused by a natural disaster, civic unrest, cyberattack, or any other threat to business operations.
+
+## Goals
+
+It is necessary to take a step back and decide what are the team's primary's objectives for this effort.
+
+### More automation
+
+More automation for the release process is desired for the team to lessen overall
+burden of our support role. These are identified areas for automation:
+
+- Release changes approval,
+- Components (API, Admin, DD-API, documentation) releases to production,
+- Lambdas releases to staging,
+- Lambdas releases to production,
+- Tests reports are published for the team to see.
+
+### Lighter release process
+
+The team desire less chores for the ops lead and incident commander, to make
+room for other tasks that cannot benefit of automation.
+
+### More release autonomy
+
+Delegate release checks to developers in order to remove burden on the ops
+lead. By having developers approve and promote their own changes ready for a
+production release, this would remove some hurdles.
+
+### Safe release to production
+
+The releases need to happen with a certain safety of checks, verifications and
+certified approvals by the change owners. Human interventions might be required
+at certain points to approve of changes or identify issues, potentially
+reverting bad rollouts. At the same time, having too many human intervention
+will lead to a higher number of mistakes.
+
+### Isolated environments for testing
+
+There are changes of certain nature such as infrastructure that leads to
+disruption of the environment itself and might require isolation in a separate
+environment.
+
+## Options
+
+This section lists suggestions made by the team with different considerations
+and goals. We'll first review the current CI/CD process.
+
+### Current process
+
+![Current CI/CD workflow](https://raw.githubusercontent.com/cds-snc/notification-adr/main/records/diagrams/current-cicd.drawio.png?raw=true "Current CI/CD workflow")
+
+### Stronger responsibility for release changes on developers
+
+A lightweight changes that the process can include in the first place is to put
+more burden on the developers' shoulders on the changes approval step.
+Specifically, whatever the developer pushes to the staging environment is what
```

Comment:
does it get released immediately? or is there a time period before the push to prod happens?

Reply:
No automation improvement on this one option, but only a slight change to unburden the ops lead.